components/proxy/src/test/kotlin/com/hotels/styx/routing/handlers/LoadBalancingGroupTest.kt

```diff
@@ -0,0 +1,218 @@
+/*
+  Copyright (C) 2013-2019 Expedia Inc.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+package com.hotels.styx.routing.handlers
+
+import com.hotels.styx.api.HttpHandler
+import com.hotels.styx.api.HttpRequest
+import com.hotels.styx.api.HttpRequest.get
+import com.hotels.styx.api.configuration.ObjectStore
+import com.hotels.styx.api.exceptions.NoAvailableHostsException
+import com.hotels.styx.routing.RoutingContext
+import com.hotels.styx.routing.RoutingObjectRecord
+import com.hotels.styx.routing.db.StyxObjectStore
+import com.hotels.styx.routing.handle
+import com.hotels.styx.routing.routingObjectDef
+import com.hotels.styx.server.HttpInterceptorContext
+import io.kotlintest.IsolationMode
+import io.kotlintest.matchers.numerics.shouldBeGreaterThan
+import io.kotlintest.matchers.types.shouldBeNull
+import io.kotlintest.shouldBe
+import io.kotlintest.shouldThrow
+import io.kotlintest.specs.FeatureSpec
+import io.mockk.mockk
+import io.mockk.verify
+import org.reactivestreams.Publisher
+import reactor.core.Disposable
+import reactor.core.publisher.toFlux
+import reactor.core.publisher.toMono
+import java.nio.charset.StandardCharsets.UTF_8
+import java.time.Duration
+
+class LoadBalancingGroupTest : FeatureSpec() {
+
+    // Tests depend on each other. So run the tests sequentially:
+    override fun isolationMode(): IsolationMode = IsolationMode.SingleInstance
+
+    init {
+        feature("Load Balancing") {
+            val factory = LoadBalancingGroup.Factory()
+            val routeDb = StyxObjectStore<RoutingObjectRecord>()
+
+            routeDb.insert("appx-01", RoutingObjectRecord.create("HostProxy", setOf("appX"), mockk(), StaticResponseHandler(200, "appx-01")))
+            routeDb.insert("appx-02", RoutingObjectRecord.create("HostProxy", setOf("appX"), mockk(), StaticResponseHandler(200, "appx-02")))
+            routeDb.insert("appx-03", RoutingObjectRecord.create("HostProxy", setOf("appX"), mockk(), StaticResponseHandler(200, "appx-03")))
+
+            routeDb.insert("appy-01", RoutingObjectRecord.create("HostProxy", setOf("appY"), mockk(), StaticResponseHandler(200, "appy-01")))
+            routeDb.insert("appy-02", RoutingObjectRecord.create("HostProxy", setOf("appY"), mockk(), StaticResponseHandler(200, "appy-02")))
+
+            routeDb.watch().waitUntil { it.entrySet().size == 5 }
+
+            val lbGroup = factory.build(listOf("appX"), RoutingContext(routeDb = routeDb).get(), routingObjectDef("""
```

Comment:
The method `RoutingContext.get` is a bit confusing. If the relationship between `RoutingContext` and `HttpHandlerFactory.Context` can't be refactored, perhaps `get` should be called `adapt` or something?

Reply:
I suppose `RoutingContext` needs to be renamed. It really is a "prototype" object for building `HttpHandlerFactory.Context`s.

However I do think that `HttpHandlerFactory`, and therefore `HttpHandlerFactory.Context` also needs a more apt name. The problem is we already have a `RoutingObjectFactory` class.

`RoutingObjectFactory` builds a `RoutingObject` from a given Yaml configuration object (`RoutingObjectConfiguration`). If the configuration defines a new object, it delegates to a particular `HttpHandlerFactory` instance to build the desired object type. If the configuration is just a named reference, it returns a name resolution function wrapped in a lambda.

`HttpHandlerFactory` is an interface. There is a separate implementation for each actual routing object type (like load balancing group, host proxy, etc).

I welcome any suggestions.

Reply:
Thinking about it, current `RoutingObjectFactory` functionality could be implemented directly in a `HttpHandlerFactory` interface as a static method.

Then we could remove the existing `RoutingObjectFactory`, and let the `HttpHandlerFactory` to take its name.

We would then have `RoutingObjectFactory.Context`, and the the Kotlin helper could be renamed as RoutingObjectFactoryContextPrototype` 😱or something along the lines. But I guess that would to be clearer?